arch/rv32i/src/syscall.rs

```diff
@@ -421,7 +626,7 @@ impl kernel::syscall::UserspaceKernelBoundary for SysCall {
                     mcause::Exception::UserEnvCall | mcause::Exception::MachineEnvCall => {
                         // Need to increment the PC so when we return we start at the correct
                         // instruction. The hardware does not do this for us.
-                        state.pc += 4;
+                        state.pc = state.pc.wrapping_add(4);
```

Comment:
Also tangential. This should be functionally equivalent on release builds, but would have panicked in debug builds. Overflowing seems like the right thing to do here.

Reply:
Thinking about this a bit more, this seems to have an explicit assumption that there is not a compressed environment call instruction. If that existed (potentially as part of a future extension), we'd need to increment by 2 bytes here instead. We should probably document this.

Reply:
This will be nightmare to debug if it hits us. Should we just fix is now.

This C code is enough to get the instruction length

```C
static inline int insn_len(uint16_t opcode)
{
    return (opcode & 3) == 3 ? 4 : 2;
}
```

So it should be pretty easy to check

Reply:
I tried to implement this, and while the logic you propose is indeed very simple, it requires us to know the opcode that our platform is current executing.

However, that's not trivial. RISC-V provides us with the `mtinst` CSR, but this CSR (where hardware implements it) will contain the transformed, expanded instruction, in case we're executing a compressed instruction. Thus we cannot use this.

Alternatively, we could try and deference the PC. However, I'm having a hard time reasoning about the safety of that -- when the application is able to do an `ecall`, presumably it has access to this instruction. But maybe there's other edge cases where its not always safe for the kerrnel to dereference the PC in this particular location, such as with Kernel MPU. And in either case, this would be an additional load & compare on the already expensive system call critical path, for a potential future instruction that does not yet exist.

Based on this, I'd rather postpone making these changes until we have an implementation where this is a problem. Notably, this will never break the kernel, it just might jump to an improper instruction offset or skip a compressed instruction in userspace.

Reply:
Interesting points. I didn't think about `mtinst` containing the expanded construction.

I suspect de-referencing the PC should be fine, but I agree about the overhead. I agree with postponing, let's see if it is ever an issue